File: loctree.mjs
#!/usr/bin/env node
import { createReadStream } from 'node:fs';
import { readdir } from 'node:fs/promises';
import path from 'node:path';

function parseArgs(argv) {
  const positional = [];
  const options = { exts: null };

  for (const arg of argv) {
    if (arg.startsWith('--ext=')) {
      const value = arg.slice('--ext='.length).trim();
      if (value) {
        options.exts = new Set(
          value
            .split(',')
            .map((segment) => segment.trim().replace(/^\./, ''))
            .filter(Boolean)
        );
      }
      continue;
    }
    if (arg.startsWith('--')) {
      console.warn(`Ignoring unknown flag ${arg}`);
      continue;
    }
    positional.push(arg);
  }

  if (positional.length === 0) {
    throw new Error('Usage: loc-tree.mjs <root> [--ext=rs,ts,tsx]');
  }

  return { root: positional[0], options };
}

function countFileLines(filePath) {
  return new Promise((resolve, reject) => {
    let count = 0;
    let sawData = false;
    let endedWithNewline = true;

    const stream = createReadStream(filePath);
    stream.on('data', (chunk) => {
      sawData = true;
      for (let i = 0; i < chunk.length; i += 1) {
        if (chunk[i] === 0x0a) {
          count += 1;
        }
      }
      endedWithNewline = chunk[chunk.length - 1] === 0x0a;
    });
    stream.on('end', () => {
      if (!endedWithNewline && sawData) {
        count += 1;
      }
      resolve(count);
    });
    stream.on('error', reject);
  });
}

async function collectLines(root, exts) {
  const lines = [];

  const sortKey = (entry) => {
    if (entry.entry.isDirectory()) {
      return `0-${entry.name.toLowerCase()}`;
    }
    return `1-${entry.name.toLowerCase()}`;
  };

  async function walk(currentPath, prefixFlags) {
    let dirEntries = await readdir(currentPath, { withFileTypes: true });
    dirEntries = dirEntries
      .filter((entry) => entry.name !== '.DS_Store')
      .map((entry) => ({ entry, name: entry.name }))
      .sort((a, b) => sortKey(a).localeCompare(sortKey(b)));

    for (let index = 0; index < dirEntries.length; index += 1) {
      const { entry, name } = dirEntries[index];
      const isLast = index === dirEntries.length - 1;
      const prefix = prefixFlags
        .map((hasNext) => (hasNext ? '│   ' : '    '))
        .join('');
      const branch = isLast ? '└── ' : '├── ';
      const label = `${prefix}${branch}${name}`;
      const fullPath = path.join(currentPath, name);
      const record = { label, loc: null };

      if (entry.isFile()) {
        const extIndex = name.lastIndexOf('.');
        const ext =
          extIndex === -1 ? '' : name.slice(extIndex + 1).trim().toLowerCase();
        if (!exts || exts.has(ext)) {
          try {
            record.loc = await countFileLines(fullPath);
          } catch (error) {
            console.warn(`Failed to count ${fullPath}: ${error.message ?? error}`);
          }
        }
      }

      lines.push(record);

      if (entry.isDirectory()) {
        await walk(fullPath, [...prefixFlags, !isLast]);
      }
    }
  }

  await walk(root, []);
  return lines;
}

async function main() {
  const { root, options } = parseArgs(process.argv.slice(2));
  const rootPath = path.resolve(root);
  const lines = await collectLines(rootPath, options.exts);

  if (lines.length === 0) {
    console.log(`${path.basename(rootPath)}/ (empty)`);
    return;
  }

  const maxLabelLen = lines.reduce(
    (max, item) => Math.max(max, item.label.length),
    0
  );

  console.log(`${path.basename(rootPath)}/`);
  for (const { label, loc } of lines) {
    if (loc == null) {
      console.log(label);
      continue;
    }
    const padding = ' '.repeat(maxLabelLen - label.length + 2);
    const locStr = loc.toString().padStart(6, ' ');
    console.log(`${label}${padding}${locStr}`);
  }
}

main().catch((error) => {
  console.error(error.message ?? error);
  process.exit(1);
});

File: loctree.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
from pathlib import Path
from typing import List, Optional, Set, Tuple


def collect_lines(root: Path, exts: Optional[Set[str]]) -> List[Tuple[str, Optional[int]]]:
    """
    Zwraca listę (label, loc):
      - label = gotowy string z gałązką drzewa + nazwą pliku/katalogu
      - loc   = liczba linii (dla plików, jeśli pasują rozszerzeniem), albo None
    """

    lines: List[Tuple[str, Optional[int]]] = []

    def sort_key(p: Path):
        # katalogi najpierw, potem pliki; wszystko po nazwie case-insensitive
        return (0 if p.is_dir() else 1, p.name.lower())

    def walk(dir_path: Path, prefix_parts: List[bool]) -> None:
        entries = sorted(list(dir_path.iterdir()), key=sort_key)
        for idx, entry in enumerate(entries):
            is_last = idx == len(entries) - 1

            # prefix z pionowymi kreskami / spacjami
            prefix = ""
            for has_more in prefix_parts:
                prefix += "│   " if has_more else "    "

            branch = "└── " if is_last else "├── "
            label = prefix + branch + entry.name

            loc: Optional[int] = None
            if entry.is_file():
                if exts is None or entry.suffix.lstrip(".") in exts:
                    try:
                        with entry.open("r", encoding="utf-8", errors="ignore") as f:
                            loc = sum(1 for _ in f)
                    except OSError:
                        loc = None

            lines.append((label, loc))

            if entry.is_dir():
                # jeśli to nie ostatni element, to w kolejnych poziomach rysujemy │
                walk(entry, prefix_parts + [not is_last])

    walk(root, [])
    return lines


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Tree z kolumną LOC dla katalogu (pół‑eza, pół‑psycho)."
    )
    parser.add_argument(
        "root",
        type=str,
        help="Katalog startowy (np. src-tauri/src)",
    )
    parser.add_argument(
        "--ext",
        type=str,
        default="",
        help="Lista rozszerzeń, po przecinku (np. 'rs,ts,tsx'). Puste = wszystkie pliki.",
    )
    args = parser.parse_args()

    root_path = Path(args.root).resolve()
    if not root_path.is_dir():
        raise SystemExit(f"{root_path} nie jest katalogiem")

    exts: Optional[Set[str]]
    if args.ext.strip():
        exts = {e.strip().lstrip(".") for e in args.ext.split(",") if e.strip()}
    else:
        exts = None

    lines = collect_lines(root_path, exts)
    if not lines:
        return

    max_label_len = max(len(label) for label, _ in lines)

    # nagłówek z nazwą root
    print(f"{root_path.name}/")

    for label, loc in lines:
        if loc is None:
            print(label)
        else:
            padding = " " * (max_label_len - len(label) + 2)
            print(f"{label}{padding}{loc:6d}")


if __name__ == "__main__":
    main()